designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,950 @@
+
+Some of the key differences from the way ESLint's configuration resolution works today are:
+
+1. There is no automatic search for `eslint.config.js` in the user's home directory. Users wanting this functionality can either pass a home directory file using `-c` or manually read in that file from their `eslint.config.js` file.
+1. Once a `eslint.config.js` file is found, there is no more searching for any further config files.
+1. There is no automatic merging of config files using either `extends` or `overrides`.
+1. When `-c` is passed on the command line, there is no search performed.
+
+### File Pattern Resolution
+
+Because there are file patterns included in `eslint.config.js`, this requires a change to how ESLint decides which files to lint. The process for determining which files to lint is:
+
+1. When a filename is passed directly (such as `eslint foo.js`):
+    1. ESLint checks to see if there is one or more configs where the `files` pattern matches the file that was passed in and does not match the `ignores` pattern. The pattern is evaluated by prepending the directory in which `eslint.config.js` was found to each pattern in `files`. All configs that match `files` and not `ignores` are merged (with the last matching config taking precedence over others). If no config is found, then the file is ignored with an appropriate warning.
+    1. If a matching config is found, then the `ignores` pattern is tested against the filename. If it's a match, then the file is ignored. Otherwise, the file is linted.
+1. When a glob pattern is passed directly (such as `eslint src/*.js`):
+    1. ESLint expands the glob pattern to get a list of files.
+    1. Each file is checked individually as in step 1.
+1. When a directory is passed directly (such as `eslint src`):
+    1. The directory is converted into a glob pattern by appending `**/*` to the directory (such as `src` becomes `src/**/*`).
+    1. The glob pattern is checked as in step 2.
+1. When a relative directory is passed directly (such as `eslint .`):
+    1. The relative directory pattern is resolved to a full directory name.
+    1. The glob pattern is checked as in step 3.
+
+**Note:** ESLint will continue to ignore `node_modules` by default.
+
+### Rename `--no-eslintrc` to `--no-config-file` and `useEslintrc` to `useConfigFile`
+
+Because the config filename has changed, it makes sense to change the command line `--no-eslintrc` flag to a more generic name, `--no-config-file` and change `CLIEngine`'s `useEslintrc` option to `useConfigFile`. In the short term, to avoid a breaking change, these pairs of names can be aliased to each other.
+
+### Implementation Details
+
+The implementation of this feature requires the following changes:
+
+1. Create a new `ConfigArray` class to manage configs.
```

Comment:
[ConfigArray](https://github.com/eslint/eslint/blob/master/lib/cli-engine/config-array/config-array.js) class exists already.
Would you create the existing `ConfigArray` instance?

The element type of config arrays is here: [/lib/cli-engine/config-array/config-array.js#L51-L69](https://github.com/eslint/eslint/blob/d55490fa73ff69416de375e4c1cd67b6edba531c/lib/cli-engine/config-array/config-array.js#L51-L69)

- `name` and `filePath` are just debugging info.
- `criteria` is a tester object to test if given path matches the `files`/`excludedFiles` fields or not. This supports AND operation for `extends` in `overrides`. Currently, the `FileEnumerator` class iterates files where match this tester if a directory path is present.
- `ignorePattern` is a tester object to test if given path matches `.eslintignore` and the `ignorePatterns` field.
- `parser` and `plugins` are loaded objects, with debugging info.
- `processor` is a string, so it needs some tweaks like https://github.com/eslint/eslint/commit/bd5de522e9c2dbeb607f4ac5c7d3d42969c03ac8.
- `env`, `globals`, `noInlineConfig`, `parserOptions`, `reportUnusedDisableDirectives`, `root`, and `settings` are the configured values as-is.
- [`extractConfig(filePath)`](https://github.com/eslint/eslint/blob/d55490fa73ff69416de375e4c1cd67b6edba531c/lib/cli-engine/config-array/config-array.js#L470-L485) method merges the array elements to a final configuration.

It will reduce the amount of code that we need to modify because we don't need to modify the use side of the config arrays mostly.

----

Currently, the creation logic of the config arrays are in [ConfigArrayFactory](https://github.com/eslint/eslint/blob/d55490fa73ff69416de375e4c1cd67b6edba531c/lib/cli-engine/config-array-factory.js) and [CascadingConfigArrayFactory](https://github.com/eslint/eslint/blob/d55490fa73ff69416de375e4c1cd67b6edba531c/lib/cli-engine/cascading-config-array-factory.js). The `ConfigArrayFactory` loads a file then creates a config array. The `CascadingConfigArrayFactory` concatenates (and caches) config arrays of all of CLI options, found config files, and `.eslintignore`.

The implementation will be alternate to those factories.

Reply:
Yes, the creation logic will be completely different from what we have today. I've renamed the class to `ESLintConfigArray` to distinguish from the current implementation (which didn't exist when I created this RFC).

Reply:
My intention was, creation logic is two existing factories and the new config system can use intermediate representation `CofnigArray` class as-is.

Reply:
Oh I see. It's unfortunate that you didn't let me know that ahead of time. I don't believe we can mix the two implementations at this point, nor do I think it would be a good idea.

Reply:
> I don't believe we can mix the two implementations at this point

I can implement it. Because the intermediate object has all of the required features to represent the new config and it's completely separated from the creation logic of the current config.

> nor do I think it would be a good idea

If the two systems have the same intermediate representation, we don't need to modify the use side of the intermediate representation. The use side doesn't need to care where the config array came from. It will simplify our codebase.